File: lantoli/ruby/roman_numerals.rb
require 'rspec'

ROMANS = { M: 1000, CM: 900, D: 500, CD: 400, C: 100, XC: 90, L: 50, XL: 40, X: 10, IX: 9, V: 5, IV: 4, I: 1 }

class Fixnum

  def to_roman
    return nil unless self > 0 && self < 4000
    remaining_number = self
    ROMANS.inject ("") do | roman_str, current_number |
        times,remaining_number = remaining_number.divmod current_number[1]
        roman_str + current_number[0].to_s * times
    end
  end

end

class String

  def from_roman
    remaining_arabic_str = self.clone
    roman_candidate = 0
    ROMANS.each do | roman, arabic |
      while (remaining_arabic_str.sub!(/^#{roman}/,'') != nil)
        roman_candidate += arabic
      end
    end
    self == roman_candidate.to_roman ? roman_candidate : nil
  end

end



TRANSFORMATIONS = {
  I: 1, II: 2, III: 3, IV: 4, V: 5, VI: 6, VII: 7, VIII: 8, IX: 9, X: 10, XI: 11, XII: 12, XIV: 14, XV: 15,
  XIX: 19, XXXIX: 39, XL: 40, XLI: 41, L: 50, LXXXIX: 89, XC: 90, XCIX: 99, C: 100, CCCXCIX: 399, CD: 400,
  D: 500, DCCCXCIX: 899, CM: 900, M: 1000, MMXI: 2011, MMMCMXCIX: 3999
}

describe "From arabic to roman numerals. " do

  TRANSFORMATIONS.each do |roman, arabic|
    it("transforms #{arabic} to #{roman}") do
      arabic.to_roman.should == roman.to_s
    end
  end

  [ -10, 0, 4000, 4100 ].each do | bad_arabic |
    it("#{bad_arabic} can not be transformed to roman numeral") do
      bad_arabic.to_roman.should == nil
    end
  end
end

describe "From roman to arabic numerals. " do

  TRANSFORMATIONS.each do |roman, arabic|
    it("transforms #{roman} to #{arabic}") do
      roman.to_s.from_roman.should == arabic
    end
  end

  [ "IIII", "VIV", "", "IA", "HOLA", "i", "v"  ].each do | bad_roman |
    it("#{bad_roman} is not a valid roman numeral") do
      bad_roman.from_roman.should == nil
    end
  end
  
end

File: cavalle/roman_numerals.js
var vows = require("vows"),
    assert = require("assert");

const ROMAN_MAPPINGS = { M:1000, CM:900, D:500, CD:400, C:100, XC:90, L:50, XL:40, X:10, IX:9, V:5, IV:4, I:1 }

Number.prototype.toRoman = function() {
  var str = "", 
      num = this.valueOf();
  
  for (var roman in ROMAN_MAPPINGS) {
    var arabic = ROMAN_MAPPINGS[roman];
    while (num >= arabic) {
      str += roman;
      num -= arabic;
    }
  }

  return str;
}

String.prototype.toArabic = function() {
  var num = 0,
      str = this.valueOf();

  for (var roman in ROMAN_MAPPINGS) {
    var arabic = ROMAN_MAPPINGS[roman];
    while (str.substr(0, roman.length) == roman) {
      num += arabic;
      str = str.substr(roman.length);
    }
  }

  return num;
}

vows.describe("Roman numerals").addBatch({
  "Number.toRoman": {
    "transforms 1 to I": function() { assert.equal(Number(1).toRoman(), "I"); },
    "transforms 2 to II": function() { assert.equal(Number(2).toRoman(), "II"); },
    "transforms 3 to III": function() { assert.equal(Number(3).toRoman(), "III"); },
    "transforms 4 to IV": function() { assert.equal(Number(4).toRoman(), "IV"); },
    "transforms 5 to V": function() { assert.equal(Number(5).toRoman(), "V"); },
    "transforms 6 to VI": function() { assert.equal(Number(6).toRoman(), "VI"); },
    "transforms 7 to VII": function() { assert.equal(Number(7).toRoman(), "VII"); },
    "transforms 8 to VIII": function() { assert.equal(Number(8).toRoman(), "VIII"); },
    "transforms 9 to IX": function() { assert.equal(Number(9).toRoman(), "IX"); },
    "transforms 10 to X": function() { assert.equal(Number(10).toRoman(), "X"); },
    "transforms 11 to XI": function() { assert.equal(Number(11).toRoman(), "XI"); },
    "transforms 39 to XXXIX": function() { assert.equal(Number(39).toRoman(), "XXXIX"); },
    "transforms 49 to XLIX": function() { assert.equal(Number(49).toRoman(), "XLIX"); },
    "transforms 89 to LXXXIX": function() { assert.equal(Number(89).toRoman(), "LXXXIX"); },
    "transforms 399 to CCCXCIX": function() { assert.equal(Number(399).toRoman(), "CCCXCIX"); },
    "transforms 499 to CDXCIX": function() { assert.equal(Number(499).toRoman(), "CDXCIX"); },
    "transforms 899 to DCCCXCIX": function() { assert.equal(Number(899).toRoman(), "DCCCXCIX"); },
    "transforms 999 to CMXCIX": function() { assert.equal(Number(999).toRoman(), "CMXCIX"); },
    "transforms 1999 to MCMXCIX": function() { assert.equal(Number(1999).toRoman(), "MCMXCIX"); }
  },
  "String.toArabic": {
    "transforms I to 1": function() { assert.equal(String("I").toArabic(), 1) },
    "transforms II to 2": function() { assert.equal(String("II").toArabic(), 2) },
    "transforms III to 3": function() { assert.equal(String("III").toArabic(), 3) },
    "transforms MCMXCIX to 1999": function() { assert.equal(String("MCMXCIX").toArabic(), 1999) },
  }
}).run();
